File: copilot-token-budget/extension/src/livebilling/labels.ts
// livebilling/labels.ts — UI labels and snapshot selection for live billing.

import { LiveBillingSnapshot, Session } from '../types';

export function liveBillingLabel(snapshot?: LiveBillingSnapshot, now: Date = new Date()): string {
  if (snapshot === undefined) {
    return '(estimated)';
  }
  if (snapshot.availability === 'unavailable') {
    return '(unavailable)';
  }
  // If SourceLabel is already set (by the refresher), use it.
  if (snapshot.sourceLabel) {
    return snapshot.sourceLabel;
  }
  // Fallback for snapshots created without explicit SourceLabel.
  return `(org aggregate, ~${hoursAgo(snapshot.lastRefreshedAt, now)}h ago)`;
}

export function latestLiveBillingSnapshot(sessions: Session[]): LiveBillingSnapshot | undefined {
  let latest: LiveBillingSnapshot | undefined;
  for (const s of sessions) {
    const snap = s.orgBillingSnapshot;
    if (snap === undefined) {
      continue;
    }
    if (latest === undefined || snap.lastRefreshedAt.getTime() > latest.lastRefreshedAt.getTime()) {
      latest = snap;
    }
  }
  return latest;
}

function hoursAgo(t: Date, now: Date): number {
  const ms = now.getTime() - t.getTime();
  if (ms <= 0) {
    return 1;
  }
  return Math.max(1, Math.floor(ms / (60 * 60 * 1000)));
}

File: copilot-token-budget/core/internal/livebilling/labels.go
package livebilling

import (
	"fmt"
	"time"
)

// DisplayLabel returns the UI label for a live billing snapshot.
// The visible contract stays deliberately small:
//   - nil or missing snapshot => (estimated)
//   - unavailable snapshot    => (unavailable)
//   - snapshot with SourceLabel => return the SourceLabel as-is
//   - otherwise               => (org aggregate, ~Xh ago)
func DisplayLabel(snapshot *OrgBillingSnapshot, now time.Time) string {
	if snapshot == nil {
		return "(estimated)"
	}
	if snapshot.Availability == AvailabilityUnavailable {
		return "(unavailable)"
	}
	// If SourceLabel is already set (by the refresher), use it.
	if snapshot.SourceLabel != "" {
		return snapshot.SourceLabel
	}
	// Fallback for snapshots created without explicit SourceLabel.
	hours := hoursAgo(snapshot.LastRefreshedAt, now)
	return fmt.Sprintf("(org aggregate, ~%dh ago)", hours)
}

func hoursAgo(t time.Time, now time.Time) int {
	if t.IsZero() {
		return 0
	}
	d := now.UTC().Sub(t.UTC())
	if d < 0 {
		d = 0
	}
	h := int(d.Hours())
	if h < 1 {
		return 1
	}
	return h
}
